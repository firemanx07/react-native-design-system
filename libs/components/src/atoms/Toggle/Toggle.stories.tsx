import { useNamespacedTheme } from '@proxym/themes';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen, StorybookSection } from '../../storybook';
import { Toggle as ToggleBase, PropsType as ToggleProps } from './Toggle';

const Story: ComponentMeta<typeof ToggleBase> = {
  component: ToggleBase,
  title: 'Atoms/Toggle',
};
export default Story;

const noop = () => {};

const Template: ComponentStory<typeof ToggleBase> = (args: ToggleProps) => {
  const { colors } = useNamespacedTheme();
  return (
    <StorybookScreen title="Toggle:" style={{ backgroundColor: colors.light }}>
      <StorybookSection>
        <ToggleBase {...args} />
      </StorybookSection>
      <StorybookSection title="inactive">
        <ToggleBase onPress={noop} isActive={false} />
      </StorybookSection>
      <StorybookSection title="inactive - secondary">
        <ToggleBase onPress={noop} isActive={false} secondary />
      </StorybookSection>
      <StorybookSection title="active">
        <ToggleBase onPress={noop} isActive />
      </StorybookSection>
      <StorybookSection title="disabled">
        <ToggleBase onPress={noop} isActive={false} disabled />
      </StorybookSection>
    </StorybookScreen>
  );
};

const parameters = {
  controls: {
    include: ['disabled', 'secondary', 'size'],
  },
};

const defaultArgs: ToggleProps = {
  isActive: false,
  disabled: false,
  secondary: false,
  size: 24,
  onPress: noop,
};

export const Toggle = Template.bind({});
Toggle.parameters = parameters;
Toggle.args = {
  ...defaultArgs,
};
