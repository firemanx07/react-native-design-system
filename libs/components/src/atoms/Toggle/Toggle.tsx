import { useNamespacedTheme, styled } from '@proxym/themes';
import React, { memo, useMemo, useCallback, useState } from 'react';
import { StyleProp, ViewStyle, Animated } from 'react-native';

import { TestIDType } from '../../types';

export type PropsType = {
  isActive: boolean;
  onPress: (isActive: boolean) => void;
  disabled?: boolean;
  secondary?: boolean;
  size?: number;
  style?: StyleProp<ViewStyle>;
} & TestIDType;

const TOGGLE_DURATION = 150;

export const Toggle = ({
  isActive: isActiveInitial = false,
  onPress,
  disabled,
  secondary,
  size,
  style,
  testID,
}: PropsType) => {
  const { iconSize, spacing, colors } = useNamespacedTheme();
  const [isActive, setIsActive] = useState<boolean>(isActiveInitial);
  const toggleSize = size || iconSize.primary;
  const toggleSpacing = spacing.tiny / 2;
  const toggleRightPosition = toggleSize - toggleSpacing * 2;
  const startPosition = isActive ? toggleRightPosition : 0;
  const endPosition = isActive ? 0 : toggleRightPosition;

  const containerWidth = toggleSize * 2;
  const containerHeight = toggleSize + toggleSpacing * 2;

  const translateX = new Animated.Value(startPosition);

  const handleToggle = useCallback(() => {
    Animated.timing(translateX, {
      toValue: endPosition,
      useNativeDriver: true,
      duration: TOGGLE_DURATION,
    }).start(() => {
      setIsActive((prev) => {
        const newActiveState = !prev;
        onPress(newActiveState);
        return newActiveState;
      });
    });
  }, [endPosition, onPress, isActive, setIsActive, translateX]);

  const toggleBackgroundColor = useMemo(() => {
    if (disabled) {
      return colors.grayLight;
    }
    if (isActive) {
      return colors.primary;
    }
    return colors.graySoft;
  }, [colors, isActive, disabled]);

  return (
    <Container
      width={containerWidth}
      height={containerHeight}
      spacing={toggleSpacing}
      backgroundColor={toggleBackgroundColor}
      onPress={handleToggle}
      style={style}
      disabled={disabled}
      activeOpacity={1}
      testID={testID}
    >
      <ToggleBullet
        size={toggleSize}
        secondary={secondary}
        style={[
          {
            transform: [{ translateX }],
          },
        ]}
      />
    </Container>
  );
};

type ContainerType = {
  width: number;
  height: number;
  spacing: number;
  backgroundColor: string;
};

const Container = styled.TouchableOpacity<ContainerType>`
  width: ${({ width }) => width}px;
  height: ${({ height }) => height}px;
  border-radius: ${({ height }) => height / 2}px;
  background-color: ${({ backgroundColor }) => backgroundColor};
  padding: ${({ spacing }) => spacing}px;
`;

type ToggleBulletType = {
  size: number;
  secondary?: boolean;
};

const ToggleBullet = styled(Animated.View)<ToggleBulletType>`
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  border-radius: ${({ size }) => size / 2}px;
  elevation: 2;
  shadow-offset: 0 2px;
  shadow-color: ${({ theme }) => theme.ds.colors.dark};
  shadow-opacity: 0.1;
  background-color: ${({ theme, secondary }) =>
    secondary ? theme.ds.colors.grayLight : theme.ds.colors.light};
`;

export default memo(Toggle);
