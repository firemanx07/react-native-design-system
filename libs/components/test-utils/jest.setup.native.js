import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();

console.warn = () => {};

Animated.timing = () => ({
  start: () => jest.fn(),
});
