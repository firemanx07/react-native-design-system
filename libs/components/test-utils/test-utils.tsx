import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
