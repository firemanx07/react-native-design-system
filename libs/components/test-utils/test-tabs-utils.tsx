import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
