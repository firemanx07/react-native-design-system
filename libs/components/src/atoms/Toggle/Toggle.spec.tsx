import { testAllStories } from '../../../test-utils';
import * as stories from './Toggle.stories';

describe('Toggle tests', () => {
  testAllStories(stories);
});
