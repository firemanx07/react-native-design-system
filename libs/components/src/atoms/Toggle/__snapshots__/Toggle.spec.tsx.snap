// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle tests Toggle snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F2F3F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
      {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <Text
    fontFamily="Muli-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#222222",
          "fontFamily": "Muli-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Toggle:
  </Text>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#DEDFE0",
          "borderRadius": 14,
          "height": 28,
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 48,
        }
      }
    >
      <View
        collapsable={false}
        secondary={false}
        size={24}
        style={
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "elevation": 2,
            "height": 24,
            "shadowColor": "#222222",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": 24,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <Text
      fontFamily="Muli-Bold"
      fontSize={16}
      lineHeight={20}
      style={
        [
          {
            "color": "#222222",
            "fontFamily": "Muli-Bold",
            "fontSize": 16,
            "lineHeight": 20,
          },
          {
            "backgroundColor": "#F2F3F4",
            "left": 16,
            "paddingHorizontal": 8,
            "position": "absolute",
            "top": -12,
          },
        ]
      }
    >
      inactive
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#DEDFE0",
          "borderRadius": 14,
          "height": 28,
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 48,
        }
      }
    >
      <View
        collapsable={false}
        size={24}
        style={
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "elevation": 2,
            "height": 24,
            "shadowColor": "#222222",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": 24,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <Text
      fontFamily="Muli-Bold"
      fontSize={16}
      lineHeight={20}
      style={
        [
          {
            "color": "#222222",
            "fontFamily": "Muli-Bold",
            "fontSize": 16,
            "lineHeight": 20,
          },
          {
            "backgroundColor": "#F2F3F4",
            "left": 16,
            "paddingHorizontal": 8,
            "position": "absolute",
            "top": -12,
          },
        ]
      }
    >
      inactive - secondary
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#DEDFE0",
          "borderRadius": 14,
          "height": 28,
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 48,
        }
      }
    >
      <View
        collapsable={false}
        secondary={true}
        size={24}
        style={
          {
            "backgroundColor": "#F2F3F4",
            "borderRadius": 12,
            "elevation": 2,
            "height": 24,
            "shadowColor": "#222222",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": 24,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <Text
      fontFamily="Muli-Bold"
      fontSize={16}
      lineHeight={20}
      style={
        [
          {
            "color": "#222222",
            "fontFamily": "Muli-Bold",
            "fontSize": 16,
            "lineHeight": 20,
          },
          {
            "backgroundColor": "#F2F3F4",
            "left": 16,
            "paddingHorizontal": 8,
            "position": "absolute",
            "top": -12,
          },
        ]
      }
    >
      active
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#FF8066",
          "borderRadius": 14,
          "height": 28,
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 48,
        }
      }
    >
      <View
        collapsable={false}
        size={24}
        style={
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "elevation": 2,
            "height": 24,
            "shadowColor": "#222222",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "transform": [
              {
                "translateX": 20,
              },
            ],
            "width": 24,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <Text
      fontFamily="Muli-Bold"
      fontSize={16}
      lineHeight={20}
      style={
        [
          {
            "color": "#222222",
            "fontFamily": "Muli-Bold",
            "fontSize": 16,
            "lineHeight": 20,
          },
          {
            "backgroundColor": "#F2F3F4",
            "left": 16,
            "paddingHorizontal": 8,
            "position": "absolute",
            "top": -12,
          },
        ]
      }
    >
      disabled
    </Text>
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#F2F3F4",
          "borderRadius": 14,
          "height": 28,
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 48,
        }
      }
    >
      <View
        collapsable={false}
        size={24}
        style={
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "elevation": 2,
            "height": 24,
            "shadowColor": "#222222",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": 24,
          }
        }
      />
    </View>
  </View>
</View>
`;
